CIP-????/README.md

```diff
@@ -0,0 +1,471 @@
+- Constructors are translated to Plutus Core `const` terms instead of using the Data `Constr` constructor.
+- Uses of the `Map` and `List` Data constructors will need more specific handling but will be translated in line with their underlying dataype representation in Plutus Core.
+
+## Rationale: how does this CIP achieve its goals?
+
+### Benefits
+
+#### 1. Avoid Decoding the Script Context
+
+This CIP solves problem 1 in the most drastic way possible: it removes it entirely.
+With the proposed ledger interface changes the script context is passed across _already decoded_, so there is zero work for the script to do.
+
+#### 2. Faster Processing of Structured Data
+
+Most Plutus Core programs spend a significant amount of their time operating on datatypes, so we would expect that a performance improvement here would make a significant difference.
+
+Indeed, this seems to be true.
+Our benchmarks[^see-appendix-3] show that this CIP leads to 0-30% real-time speedups depending on the program (the 0% is a primality tester that is mostly doing arithmetic, the 30% is mostly doing datatype manipulation). This is a very significant improvement, and should result in a similar reduction in budget usage.
+
+[^see-appendix-3]: See Appendix 3 for the table of results.
+
+#### 3. Simple Lifting and Unlifting
+
+Today, lifting is complicated, and unlifting is mostly impossible.
+With this CIP lifting becomes very simple, and unlifting becomes not only possible but simple.
+Our prototype implementation contains code-generation to generate lifting and unlifting code in Haskell, but it is straightforward enough that you could easily do it in any language.
+
+There are still some limitations, e.g. you cannot lift or unlift structures that contain functions, but this is not that unusual (similarly, you typically can't serialise such structures).
+
+### Costs
+
+The costs of this proposal are substantial.
+
+Plutus Core today has only 8 kinds of term, this proposal adds 2 more, an increase of 25%.
+Additionally, the new term types are the first term types which have a variable number of children.
+Both of these changes increase implementation complexity throughout the system: everything that processes Plutus Core terms must now handle the new cases, and handle the variable numbers of children.
+
+This change also modestly expends our novelty budget.
+Plutus Core is a deliberately conservative language: for the most part it is just the untyped lambda calculus.
+The proposed new features for sums-of-products are also conservative, and follow a very typical pattern.
+But they are not _quite_ as standard as the untyped lambda calculus.
+
+Moreover, a change like this would be very painful to reverse, so we should seriously consider the costs before proceeding.
+
+### Discussion
+
+#### Why is case not strict?
+
+This makes it different to all other Plutus Core terms, but we believe it is the expected behaviour: _no_ language has strict case destructuring.
```

Comment:
I feel like this opens up an interesting potential meta-issue. Currently we don't support basic short-circuiting boolean operations like `&&` and `||`. If somebody wanted to get those, they'd have to use a separate toolchain for generating UPLC code and that toolchain would need to elaborate an expression like `x || y` to something like

```
force (ifThenElse x (delay True) (delay y))
```

which is pretty complex and is awkward to integrate into the surface language in a seamless way, particularly if you're aiming for the general case. And isn't zero-cost!

However elaborating `x || y` to 

```
case x of
  True  -> True
  False -> y
```

is fast and much more convenient and begs for being included in a DSL or even a standalone language with macros or two-stage evaluation or whatever. It might become hard for us to compete with that!

Realistically though, if it becomes a problem we too can special-case `&&` and `||` and it's less likely someone will need lazy semantics for other stuff. Although in the blockchain context where every bit matters it's generally a good idea to choose tools giving you more control over performance and we've seen how far people are willing to go to improve upon costs.

Reply:
> However elaborating x || y to <case expression> is fast and much more convenient and begs for being included in a DSL or even a standalone language with macros or two-stage evaluation or whatever.

Note that even today elaborating it to the case expression will get evaluated lazily, just with more overhead. The problem is that our function calls are strict. Indeed, a macro system or something would fix that (and indeed you can easily do that today with some TH if you actually care :trollface: ). I'm unsure how big a deal it is...